### PR TITLE
Rewrite createTextWriter to buffer appends

### PR DESCRIPTION
**Pull request by amcasey**
A more ambitious version of https://github.com/microsoft/TypeScript/pull/44241

### CODE REVIEW

**Comment by amcasey** [comment]
@typescript-bot perf test this

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -3953,11 +3973,10 @@ namespace ts {
         function writeText(s: string) {
             if (s && s.length) {
                 if (lineStart) {
-                    s = getIndentString(indent) + s;
+                    fragments.push(getIndentString(indent));
```
Comment (amcasey) [inline src/compiler/utilities.ts:3976]: Forgot to update length here.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -3996,16 +4016,32 @@ namespace ts {
 
         function writeLine(force?: boolean) {
             if (!lineStart || force) {
-                output += newLine;
-                lineCount++;
-                linePos = output.length;
-                lineStart = true;
+                appendAndUpdateLineCountAndPos(newLine);
```
Comment (amcasey) [inline src/compiler/utilities.ts:4019]: Bonus failure: the tests pass empty string as `newLine` and depend on this to mark the next position as a line-start regardless.

**Comment by amcasey** [comment]
@typescript-bot perf test this

**Comment by amcasey** [comment]
This appears to have cut memory usage by ~10% in the target internal project.

**Comment by amcasey** [comment]
I switched back from `join` to concatenation and the allocation win went away.

**Comment by dmichon-msft** [comment]
How's the perf if you only replace the computeLineStarts() call with a scan for count and last line start position?

**Comment by amcasey** [comment]
Optimizing the slow case (i.e. when you need to do line math) isn't very worthwhile since it's so rare.

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Comment by amcasey** [comment]
Markdown test

<details>
<summary>Successfully analyzed 1 of 3 visited repos</summary>

| Outcome | Count |
|---------|-------|
| PackageInstallFailed | 2 |
| NewBuildHadErrors | 1 |

</details>

**Comment by amcasey** [comment]
@typescript-bot user test this
@typescript-bot test top100

**Comment by amcasey** [comment]
@typescript-bot user test tsserver
@typescript-bot test tsserver top100

**Comment by amcasey** [comment]
@typescript-bot user test tsserver
@typescript-bot test tsserver top100

**Comment by amcasey** [comment]
@typescript-bot test tsserver top100

**Comment by jakebailey** [comment]
Closing in favor of #55657.

**Pull request closed without merging**